hack/generator/pkg/conversions/property_conversions.go

```diff
@@ -971,6 +971,13 @@ func copyKnownType(name astmodel.TypeName, methodName string) func(*TypedConvers
 		}
 
 		return func(reader dst.Expr, writer func(dst.Expr) []dst.Stmt, generationContext *astmodel.CodeGenerationContext) []dst.Stmt {
+			if methodName == "DeepCopy" {
```

Comment:
Open to a cleaner way to do this - a parameter to the function indicating if we should dereference or not?

Reply:
I don't think this will generate the right code. 

For one, need to store the value in a local to dereference it, and for another if this is being assigned to an optional property, it will be wrapped in AddressOf (`&`) giving some really odd results.

Reply:
We don't need to store the value in a local to dereference it -- you just need a local for making a pointer (not for dereferencing one), see: https://play.golang.org/p/0Oc53WcpnJr

I noted below in the comment: 
> // This dereference is always safe because we ensured that both source and destination are always
> // non optional

Is this not correct? Just because it _looks required_ here doesn't necessarily mean the final destination is required?

In the case of VMSS, here's the type in question that has the `v1.JSON`:
```
//Generated from: https://schema.management.azure.com/schemas/2020-12-01/Microsoft.Compute.Extensions.json#/definitions/genericExtension
type GenericExtension struct {
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:MinLength=1
	//Publisher: Microsoft.Compute/extensions - Publisher
	Publisher string `json:"publisher"`

	// +kubebuilder:validation:Required
	//Settings: Microsoft.Compute/extensions - Settings
	Settings map[string]v1.JSON `json:"settings"`

	// +kubebuilder:validation:Required
	// +kubebuilder:validation:MinLength=1
	//Type: Microsoft.Compute/extensions - Type
	Type string `json:"type"`

	// +kubebuilder:validation:Required
	// +kubebuilder:validation:MinLength=1
	//TypeHandlerVersion: Microsoft.Compute/extensions - Type handler version
	TypeHandlerVersion string `json:"typeHandlerVersion"`
}
```

and here's the generated method that this code helps make:
```
        // Settings
	settingMap := make(map[string]v1.JSON)
	for settingKey, settingValue := range genericExtension.Settings {
		// Shadow the loop variable to avoid aliasing
		settingValue := settingValue
		settingMap[settingKey] = *settingValue.DeepCopy()
	}
	destination.Settings = settingMap
```

Is the real issue here that when dealing with a `v1.JSON` it should end up as `*v1.JSON` on the storage type?

Reply:
The conversion pipeline works by unwrapping/simplifying endpoints - so optional properties are simplified to required properties, and then rewrapped once the underlying code for the required value is generated.

The required invariant is that the `reader` and `writer` have to be generated code that matches the type specified of the matching endpoint. I _think_ you have maintained that, but will add some tests to verify.

Reply:
Resolving this as you did add some tests and fixed some corner cases that worked but were ugly.